test/unit/org/apache/cassandra/transport/NativeTransportTimeoutTest.java

```diff
@@ -116,18 +167,80 @@ public void testAsyncStageLoadShedding() throws Throwable
+                       targetClass = "org.apache.cassandra.transport.Message",
+                       targetMethod = "getCustomPayload",
+                       targetLocation = "AT ENTRY",
+                       condition = "flagged(Thread.currentThread())",
+                       action = "clear(Thread.currentThread()); " +
+                                "return org.apache.cassandra.transport.NativeTransportTimeoutTest.CUSTOM_PAYLOAD;") })
+    public void testAsyncStageLoadSheddingWithCndbStartTime() throws Throwable
+    {
+        CassandraRelevantProperties.NATIVE_TRANSPORT_ASYNC_READ_WRITE_ENABLED.setBoolean(true);
+        long nativeTransportTimeoutNanos = 10_000_000_000L;
+        DatabaseDescriptor.setNativeTransportTimeout(nativeTransportTimeoutNanos, TimeUnit.NANOSECONDS);
+
+        try
+        {
+            createTable("CREATE TABLE %s (pk int PRIMARY KEY, v text)");
+
+            Statement statement = new SimpleStatement("SELECT * FROM " + KEYSPACE + '.' + currentTable());
+
+            // rewind the start time by the native transport timeout to ensure the timeout is exceeded
+            // note that we cannot set the payload in the statement directly, because that would cause the sync stage, which
+            // precedes the async stage, to timeout
+            long mockedStartTimeNanos = MonotonicClock.approxTime.now() - nativeTransportTimeoutNanos;
+            long mockedStartTimeMillis = TimeUnit.MILLISECONDS.convert(mockedStartTimeNanos, TimeUnit.NANOSECONDS);
+            CUSTOM_PAYLOAD = Collections.singletonMap("CNDB_START_TIME", ByteBufferUtil.bytes(mockedStartTimeMillis));
+
+            doTestLoadShedding(true, statement, false);
+
+            Statement insert1 = new SimpleStatement("INSERT INTO " + KEYSPACE + '.' + currentTable() + " (pk, v) VALUES (1, 'foo')");
+            Statement insert2 = new SimpleStatement("INSERT INTO " + KEYSPACE + '.' + currentTable() + " (pk, v) VALUES (2, 'bar')");
+            statement = new BatchStatement().add(insert1).add(insert2);
+            doTestLoadShedding(true, statement, false);
+
+            PreparedStatement ps = sessionNet().prepare("SELECT * FROM " + KEYSPACE + '.' + currentTable());
+            doTestLoadShedding(true, ps.bind(), false);
+        }
+        finally
+        {
+            CassandraRelevantProperties.NATIVE_TRANSPORT_ASYNC_READ_WRITE_ENABLED.setBoolean(false);
+        }
+    }
+
     private void doTestLoadShedding(boolean useAsyncStages, Statement statement) throws InterruptedException
     {
-        EXECUTE_BARRIER = new Semaphore(0);
-        WAIT_BARRIER = new Semaphore(0);
+        doTestLoadShedding(useAsyncStages, statement, true);
+    }
+
+    private void doTestLoadShedding(boolean useAsyncStages, Statement statement, boolean injectDelay) throws InterruptedException
+    {
+        // if we're injecting a delay, we need to synchronize the execution of the statement with the test
+        if (injectDelay)
```

Comment:
In principle, we can still inject/synchronize the new tests, I just wanted to make it explicit that since we control the start time it self (rather than the timeout itself) we don't need to delay the request, even if we do, we can just rewind start time with an extra 10ms as other tests do. 

The tests pass consistently locally, if needed, I can still subtract the 2 * approxTime.error() portion of the delay

Reply:
> The tests pass consistently locally, if needed, I can still subtract the 2 * approxTime.error() portion of the delay

Adjusted timeouts with error correction because after switching to REQUEST_CREATE_NANOS 1 test become flaky.

Reply:
I need to have another look tom at the tests and make sure they run consistently, especially after switch to epoch (previously I was mapping the clock nanos but for proper testing, I need to mimic what the router would send, which is epoch and translate to clock. With that in mind, I'm not sure yet how to apply error correction)

Reply:
I settled on the existing approach, the 10ms artificial and error correction combo seems the most reliable (and consist with other tests) way to run https://github.com/datastax/cassandra/pull/1672/commits/cb57e17e591575b66341ce917e935dd3963f2fe2